src/memory_manager.rs

```diff
@@ -419,6 +665,143 @@ impl<M: Memory> MemoryManagerInner<M> {
         // Ceiling division.
         (num_pages + self.bucket_size_in_pages as u64 - 1) / self.bucket_size_in_pages as u64
     }
+
+    fn free(&mut self, id: MemoryId) {
+        self.memory_sizes_in_pages[id.0 as usize] = 0;
+        let buckets = self.memory_buckets.remove(&id);
+        if let Some(vec_buckets) = buckets {
+            for bucket in vec_buckets {
+                self.freed_buckets.push_back(bucket);
+                self.allocated_buckets -= 1;
+            }
```

Comment:
As discussed before, we should reuse freed buckets  in sorted order (i.e. freed bucket `n` should be used before freed bucket `n - 1`). Changing `push_front` to `push_back` doesn't change that. Maybe using a BTreeSet would be more appropriate here.

Reply:
Done.